[RFC] Remove unused const FEAT_GUI*

Hettomei opened the pull request:
These variables are no more needed so I removed it.

And also it simplify logic for mch_errmsg and mch_msg because we can group UNIX code.

Commit:
Macro cleanup: FEAT_GUI and ALWAYS_USE_GUI

Pyrohh commented on src/nvim/message.c:
```diff
@@ -2234,15 +2234,11 @@ void mch_errmsg(char *str)
-    isatty(2)
-# endif
-    ) {
+  if (isatty(2)) {
```
Should probably use `os_isatty`.

Hettomei replied:
You said I just have to replace every ```isatty``` with os_isatty without any other check ?

Pyrohh replied:
> without any other check ?

What do you mean? I just mean replace the two instances of `isatty` in this PR with `os_isatty`, since they have the same semantics AFAICT:
```
/// Test whether a file descriptor refers to a terminal.
///
/// @param fd File descriptor.
/// @return `true` if file descriptor refers to a terminal.
bool os_isatty(int fd)
{
    return uv_guess_handle(fd) == UV_TTY;
}

```

Pyrohh replied:
Sorry if that came across as overly blunt, I'm just trying to be clear.

Hettomei replied:
By "without any other check" I mean 'I don't now what really do ```os_isatty``` instead of ```isatty``` so I asked if I really can do this without any other check.

isatty replaced in commit a083983

Pyrohh commented:
Besides the nit, LGTM.

yagebu commented:
While greping the source for FEAT_GUI, I found some more unused `FEAT_GUI_XYZ`, that can safely be removed (namely `FEAT_GUI_GTK`, `FEAT_GUI_MAC`, `FEAT_GUI_MOTIF`, `FEAT_GUI_MSWIN`, `FEAT_GUI_W32`). Feel free to to remove them in this PR, or I can open a separate one if you want.

Hettomei commented:
Thank you. I will do that in this PR as soon as possible

Commit:
Macro cleanup: FEAT_GUI_W32

Hettomei commented:
I change this PR as WIP because I found lot of const no more used and I want to clean it up.

Commit:
Macro cleanup: FEAT_GUI_MOTIF

Commit:
Macro cleanup: USE_FILE_CHOOSER

Commit:
Macro cleanup: FEAT_BEVAL

Commit:
Macro cleanup: FEAT_TOOLBAR

Commit:
Macro cleanup: FEAT_GUI_GTK

Commit:
Macro cleanup: FEAT_GUI_X11

Commit:
Macro cleanup: FEAT_GUI_MAC

Commit:
Macro cleanup: FEAT_GUI_MSWIN

Commit:
Macro cleanup: FEAT_BEVAL_TIP

Hettomei commented:
I think it is good. 
Off course there is other unused const but I will continue to clean these in a new PR.

Change from WIP to RFC.

Pyrohh commented:
Could you prepend `Macro cleanup: ` before each commit, to match previous PRs like #2011 and #1816.

Pyrohh commented on runtime/doc/debugger.txt:
```diff
@@ -119,7 +119,6 @@ The following compile time preprocessor variables control the features:
     Debug Glyphs				FEAT_SIGNS
     Debug Highlights				FEAT_SIGNS
     Message Footer				FEAT_FOOTER
-    Balloon Evaluation				FEAT_BEVAL
```
There's still a lot of references in the docs to this, but that should probably be dealt with elsewhere so I suppose it's fine.

Hettomei replied:
Yes, about these const : FEAT_SIGNS, FEAT_FOOTER and FEAT_BEVAL,  they only appears in this doc. 

I wasn't sure if I can delete all the ```2. Vim Compile Options					*debugger-compilation*``` paragraph . Should I do this in a commit ?

Pyrohh replied:
> I wasn't sure if I can delete all the 2. Vim Compile Options *debugger-compilation* paragraph . Should I do this in a commit ?

Good idea, but do you mind If instead I do that in #2532? That PR is mostly about documentation while this one already has a pretty specific focus (cleaning up macros). I'll be sure to credit your in the commit message, but it's up to you.

Pyrohh replied:
And if you do decide to do it here, make sure to remove section 3 as well: it's empty.

Hettomei replied:
Ok, I let you do this. And don't credit me, I just revealed a probably dead doc.

Pyrohh replied:
> And don't credit me, I just revealed a probably dead doc.

I just put `spotted by @Hettomei`, that's all :-)

Pyrohh commented:
Besides the nits on the commit structure, LGTM. Good thing you mentioned those defines.

Hettomei commented:
I rebased on master and squashed/changed logs according to comments.

I have done a ```push --force``` is that a good way to do a PR ?

Pyrohh commented:
> I have done a push --force is that a good way to do a PR ?

Yes, that's fine.

Pyrohh commented:
I'll just add a waiting label so this isn't forgotten.

glts commented on src/nvim/menu.c:
```diff
@@ -1360,61 +1329,6 @@ void ex_emenu(exarg_T *eap)
-        if (menu->children == NULL) {
-          /* found a menu item instead of a sub-menu */
-          if (*p == NUL)
-            EMSG(_("E336: Menu path must lead to a sub-menu"));
```
Now that message E336 is gone you can delete the E336 messages in `src/nvim/po/*.po`. And perhaps also the tag `*E336*` in `runtime/doc/gui.txt`.

Pyrohh replied:
My understanding is that they they shouldn't be deleted as they're used for generating fuzzy translations at compile time. Not completely sure though, cc @fwalch.

Hettomei replied:
I was about to commit right now... I wait.

Also there is E338 which is never found in source.

fwalch replied:
Yup, translations shouldn't be deleted from the PO files, even if they're unused.

glts replied:
@fwalch Didn't know that, thanks. Very sorry for the noise, @Hettomei.

Hettomei replied:
Thank you for the insight.

I m newbie to C, never used po files and english is not my native language so : what concretly mean "might be useful for fuzzy translations" ?

fwalch replied:
Fuzzy translations (as opposed to "manual" translations done by a human) are automatic translations, where an existing translation is used to translate a similar, but different message. Obviously it won't be an exact translation, hence "fuzzy".

Pyrohh replied:
Great, now that that's been sorted are you ok with this PR @glts?

glts commented on src/nvim/menu.c:
```diff
@@ -1360,61 +1329,6 @@ void ex_emenu(exarg_T *eap)
-  }
-
-  if (menu == NULL)
-    EMSG(_("E337: Menu not found - check menu names"));
```
Same for E337.

glts commented on src/nvim/buffer.c:
```diff
@@ -3570,20 +3570,6 @@ void fname_expand(buf_T *buf, char_u **ffname, char_u **sfname)
     *sfname = *ffname;
   *ffname = (char_u *)fix_fname((char *)*ffname);     /* expand to full path */
-
-#ifdef FEAT_SHORTCUT
```
`FEAT_SHORTCUT` seems to be used for Windows in legacy Vim. Can we be sure that we won't need this feature?

Pyrohh replied:
I think it should be fine. I just found #460.

glts replied:
Very useful info. @Hettomei If you happen to touch the relevant commit again, please add a `References #460.` statement in the commit message.

justinmk replied:
How about changing the check to WIN32 and leave the code?

glts commented on src/nvim/eval.c:
```diff
@@ -9911,9 +9911,7 @@ static void f_has(typval_T *argvars, typval_T *rettv)
-#if defined(FEAT_BROWSE) && (defined(USE_FILE_CHOOSER) \
-    || defined(FEAT_GUI_W32) \
-    || defined(FEAT_GUI_MOTIF))
+#if defined(FEAT_BROWSE)
```
You could consider shortening this to `#ifdef FEAT_BROWSE` now.

glts commented on src/nvim/message.c:
```diff
@@ -2234,15 +2234,11 @@ void mch_errmsg(char *str)
   int len;
 
-#if (defined(UNIX) || defined(FEAT_GUI)) && !defined(ALWAYS_USE_GUI)
+#if defined(UNIX)
```
Could be `#ifdef` now.

glts commented on src/nvim/message.c:
```diff
@@ -2284,16 +2280,12 @@ void mch_errmsg(char *str)
 void mch_msg(char *str)
 {
-#if (defined(UNIX) || defined(FEAT_GUI)) && !defined(ALWAYS_USE_GUI)
+#if defined(UNIX)
```
same

glts commented on src/nvim/vim.h:
```diff
@@ -338,14 +338,9 @@ enum {
-  && !defined(FEAT_GUI_GTK) && !defined(FEAT_GUI_MAC)
 # define mch_errmsg(str)        fprintf(stderr, "%s", (str))
 # define display_errors()       fflush(stderr)
 # define mch_msg(str)           printf("%s", (str))
```
You could consider removing the space before `define` now that it's not nested any longer: `#define` ...

glts commented on src/nvim/vim.h:
```diff
@@ -338,14 +338,9 @@ enum {
 # define display_errors()       fflush(stderr)
 # define mch_msg(str)           printf("%s", (str))
-#else
-# define USE_MCH_ERRMSG
```
Just in case you hadn't noticed: `USE_MCH_ERRMSG` is now no longer defined in in the source code -- only in `scripts/stripdecls.py`. I suppose that's correct, right?

Hettomei replied:
```USE_MCH_ERRMSG```  is referenced in files ```src/nvim/globals.h``` and ```src/nvim/message.c```

And because of ```scripts/stripdecls.py``` I thought an user can add or remove this const at compile time by adding an argument, so i didn't deleted it. 

I didn't understand the purpose of scripts/stripdecls.py and so I wouldn't remove this const until anybody say "Yes you can fully delete it"

glts replied:
I'm fairly confident now that `USE_MCH_ERRMSG` can be removed.

The one instance in `message.c` used to be for error output in some GUIs when they didn't have their native error output channel ready yet during startup. This then either used `fprintf` to stderr for the message or collected the messages in `garray_T error_ga`. `error_ga` in `globals.h` is the other instance guarded by `USE_MCH_ERRMSG`. I believe @ZyX-I's `stripdecls.py` script was a one time thing, so we can just leave it there.

A quick web search confirms that `USE_MCH_ERRMSG` isn't used outside of the Vim source code.

justinmk replied:
Correct on all accounts. 

> This then either used fprintf to stderr for the message or collected the messages in garray_T error_ga.

We do still need that logic though. Probably we should forget the printf() stuff and always just collect the error messages, then decide what to do with them later.

glts commented:
Nice improvement :+1:

justinmk commented on src/nvim/eval.c:
```diff
@@ -12423,17 +12421,6 @@ static void f_resolve(typval_T *argvars, typval_T *rettv)
 #endif
 
   p = get_tv_string(&argvars[0]);
-#ifdef FEAT_SHORTCUT
```
WIN32

Hettomei commented:
I've done : 
+ Add References #460 to commit
+ replace if defined with ifdef
+ remove nested space for define

and push --force

Waiting for new comments before merge ;)

Pyrohh commented:
@Hettomei I saw you rebased, did you see @justinmk's message regarding `WIN32`? I think it might be a good idea to do what he mentioned instead of removing the code, as the code doesn't seem to be completely understood.

Hettomei commented:
Sorry, I read the message and forgot :/

I will replace FEAT_SHORTCUT with WIN32

Commit:
Macro cleanup: FEAT_SHORTCUT, replace with WIN32

Hettomei commented:
@justinmk replaced FEAT_SHORTCUT with WIN32

if I follow your idea, every single macro only defined in vim-source ```/src/os_win32.h``` should be replaced with WIN32 in nvim source to reduce the number of macro and ensure compatibility ?

The other one would be SYNC_DUP_CLOSE wich was only defined for 
src/os_msdos.h
src/os_win16.h
src/os_win32.h

Should I update ```SYNC_DUP_CLOSE``` from ```src/nvim/memfile.c``` with ```WIN32``` ?

justinmk commented:
> every single macro only defined in vim-source /src/os_win32.h should be replaced with WIN32 in nvim source to reduce the number of macro and ensure compatibility ?

I can't really say, and don't have time to look at every single macro :)

Also, WIN32 was a hint, not necessarily the exact answer. I have seen __WIN32__, _WIN32, and WIN3264. Maybe @equalsraf has arrived at a pattern in his mega-PR, would like to hear his opinion.

> Should I update SYNC_DUP_CLOSE from src/nvim/memfile.c with WIN32 ?

Maybe, but I'd rather wait until someone has time to more closely analyze that function and see if it is replaceable with libuv calls instead of us fiddling with fsync.

equalsraf commented:
As far as CMake is [concerned](http://www.cmake.org/Wiki/CMake_Useful_Variables) it is WIN32 for anything built in windows, **regardless of compiler/runtime**.

> every single macro only defined in vim-source /src/os_win32.h should be replaced with WIN32 in nvim source to reduce the number of macro and ensure compatibility ?

In theory yes, however `mch_resolve_shortcut()` is gone from the tree, this would fail to build until someone re-implements it. Can we keep the block disabled until someone takes the time to to fix it? A comment would be enough to known it means WIN32.

SYNC_DUP_CLOSE is a complete mystery to me - Mingw most likely provides fsync(). MSVC has [_commit](https://msdn.microsoft.com/en-us/library/17618685.aspx). I'd say a FIXME and an issue would be best.

Commit:
Temporary disable code which references mch_resolve_shortcut

This function is only used for windows
'mch_resolve_shortcut' can be found in vim source src/os_mswin.c
-> mch_resolve_shortcut(char_u *fname)

Spotted by @equalsraf, see PR #2609

Hettomei commented:
@equalsraf I made a commit regarding your last comment , see dbca190c16962b336787bd59137d99abc70058f8

equalsraf commented:
@Hettomei that works :+1:

Pyrohh commented:
@justinmk is there any reason you're assigned to this? I noticed a few days ago you assigned yourself to a bunch of weird issues, with #301 being of note :laughing:

I'm not sure, but https://github.com/neovim/neovim/commit/943c7479c79c156463464ec6d10929f6cef57c45 and all the instances of `#[number]` might have something to do with it.

justinmk commented:
> I'm not sure, but 943c747 and all the instances of #[number] might have something to do with it.

Yep, that's exactly right. That commit may have erroneously closed some issues.

Pyrohh commented:
> Yep, that's exactly right. That commit may have erroneously closed some issues.

Sounds like it. Do you mind if I deal with all the redundant assignments, or were you going to do it?

I commented in this PR specifically because the assignee field might made me think you're about to merge this (and I wanted to give this PR another review).

justinmk commented:
> Do you mind if I deal with all the redundant assignments, or were you going to do it?

That would be great. Only numbers in 943c747 are wrong though... I assigned some other weird ones for other reasons.

Pyrohh commented on src/nvim/buffer.c:
```diff
@@ -3576,19 +3576,21 @@ void fname_expand(buf_T *buf, char_u **ffname, char_u **sfname)
-    }
-  }
-#endif
+// TODO(Hettomei,equalsraf): Temporary disable this block
```
To reduce the diff's size, maybe just `#if 0`?

Pyrohh commented on src/nvim/eval.c:
```diff
@@ -12419,16 +12417,18 @@ static void f_resolve(typval_T *argvars, typval_T *rettv)
-    else
-      rettv->vval.v_string = vim_strsave(p);
-  }
+#ifdef WIN32
```
same here regarding #if 0

Pyrohh commented on src/nvim/vim.h:
```diff
@@ -332,20 +332,15 @@ enum {
 /* This has to go after the include of proto.h, as proto/gui.pro declares
  * functions of these names. The declarations would break if the defines had
  * been seen at that stage.  But it must be before globals.h, where error_ga
  * is declared. */
```
This comment can likely be deleted. `proto.h` doesn't exist anymore.

Hettomei replied:
@Pyrohh Is it a comment for you or should I transform this comment to 

```
/* It must be before globals.h, where error_ga is declared. */
```

(even if not related with this PR)

justinmk commented:
@Hettomei could you rebase this and remove https://github.com/Hettomei/neovim/commit/6672bc42297c34a7c347240e79e2597bda5c4c81 (the last commit)? I don't really understand the purpose of it, we're going to be working on win32 support in preparation for 0.1, and that feature simply needs to be implemented.

justinmk commented:
Merged without https://github.com/Hettomei/neovim/commit/6672bc42297c34a7c347240e79e2597bda5c4c81

Thanks

Hettomei commented:
Hi,

I'm sorry I came too late to do what you have asked.

I'm glad this was merged, again, thank you for all your hard works.
 
I will send better PR with better explanation for you to have less work to do ;)
